public/app/features/logs/logsFrame.ts

```diff
@@ -35,14 +35,32 @@ const DATAPLANE_SEVERITY_NAME = 'severity';
 const DATAPLANE_ID_NAME = 'id';
 const DATAPLANE_LABELS_NAME = 'labels';
 
+// NOTE: this is a hot fn, we need to avoid allocating new objects here
 export function logFrameLabelsToLabels(logFrameLabels: LogFrameLabels): Labels {
-  const result: Labels = {};
+  let needsSerialization = false;
 
-  Object.entries(logFrameLabels).forEach(([k, v]) => {
-    result[k] = typeof v === 'string' ? v : JSON.stringify(v);
-  });
+  for (const k in logFrameLabels) {
+    const v = logFrameLabels[k];
 
-  return result;
+    if (typeof v !== 'string') {
+      needsSerialization = true;
+      break;
```

Comment:
```suggestion
```
I think this one can be skipped.

Reply:
why would we keep looping over other labels once we found any that triggers the condition?

Reply:
Nah, I tripped out with the diff. I saw the `if (needsSerialization)` inside of the previous for loop.